[GlobalISel] fdiv to fmul transform

davemgreen opened the pull request:
This is a port of the SDAG DAGCombiner::combineRepeatedFPDivisors combine that looks for multiple fdiv operations with the same divisor and converts them to a single reciprocal fdiv and multiple fmuls. It is currently a fairly faithful port, with some additions to make sure that the newly created fdiv dominates all new uses. Compared to the SDAG version it also drops some logic about splat uses which assumes no vector fdivs and some logic about x/sqrt(x) which does not yet apply to GISel.

jayfoad commented:
> This is a port of the SDAG DAGCombiner::combineRepeatedFPDivisors combine that looks like multiple fdiv operations

Is "looks like" a typo?

aemerson commented on llvm/test/CodeGen/AArch64/fdiv-combine.ll:
```diff
@@ -135,26 +103,47 @@ define void @two_fdiv_double(double %D, double %a, double %b) #0 {
+; CHECK-GI-NEXT:    fmul s2, s3, s5
+; CHECK-GI-NEXT:    fmov s3, s4
+; CHECK-GI-NEXT:    b foo_4f
+  %div = fdiv float %a, %D
+  %div1 = fdiv float %b, %D
+  %div2 = fdiv float %c, %D
+  %div3 = fdiv float %D, %D
```
Why can't we match SDAG here with the two %D operands?

davemgreen replied:
This test was mostly added to make sure it wouldn't fail. I don't think I would expect a fdiv x / x with some fast-math flags to survive to this point.

davemgreen committed:
[GlobalISel] fdiv to fmul transform

This is a port of the SDAG DAGCombiner::combineRepeatedFPDivisors combine that
looks like multiple fdiv operations with the same divisor and converts them to
a single reciprocal fdiv and multiple fmuls. It is currently a fairly faithful
port, with some additions to make sure that the newly created fdiv dominates
all new uses. Compared to the SDAG version it also drops some logic about splat
uses which assumes no vector fdivs and some logic about x/sqrt(x) which does
not yet apply to GISel.
